source/dub/dependencyresolver.d

```diff
@@ -54,236 +54,149 @@ class DependencyResolver(CONFIGS, CONFIG) {
-						if (config == CONFIG.invalid && ch.depType != DependencyType.required)
-							continue;
-
-						// if we are at the root level, we can safely skip the maxcpi computation and instead choose another childidx config
-						if (parentbase == root_base_pack) {
-							error = ConflictError(ConflictError.Kind.noRootMatch, parent, ch, config);
-							return childidx;
-						}
-
-						if (childidx > maxcpi) {
-							maxcpi = max(childidx, parentidx);
-							error = ConflictError(ConflictError.Kind.childMismatch, parent, ch, config);
-							logDebug("%s (ci=%s)", error, maxcpi);
-						}
-
-						// we know that either the child or the parent needs to be switched
-						// to another configuration, no need to continue with other children
-						if (config == CONFIG.invalid) break;
-					}
-
-					maxcpi = max(maxcpi, validateConfigs(chnode, error));
+			// restrain the configurations to the current dependency spec
+			bool any_config = false;
+			foreach (i, c; di.configs)
+				if (di.included[i]) {
+					if (!matches(dep.configs, c))
+						di.included[i] = false;
+					else any_config = true;
 				}
+
+			if (!any_config && dep.depType == DependencyType.required) {
+				if (di.configs.length)
+					throw new ResolveException(n, dep);
+				else throw new DependencyLoadException(n, dep);
 			}
-			return maxcpi;
 		}
 
-		Nullable!ConflictError first_error;
-		size_t loop_counter = 0;
-
-		// Leave the possibility to opt-out from the loop limit
-		import std.process : environment;
-		bool no_loop_limit = environment.get("DUB_NO_RESOLVE_LIMIT") !is null;
+		try constrainDependencies(n, dependencies, 0, packs, loop_counter);
+		catch (ResolveException e) {
+			// add any relevant information for this stack frame
+			e.chain(n, dependencies);
+			throw e;
+		}
+	}
 
-		while (true) {
-			assert(no_loop_limit || loop_counter++ < 1_000_000,
-				"The dependency resolution process is taking too long. The"
-				~ " dependency graph is likely hitting a pathological case in"
-				~ " the resolution algorithm. Please file a bug report at"
-				~ " https://github.com/dlang/dub/issues and mention the package"
-				~ " recipe that reproduces this error.");
-
-			// check if the current combination of configurations works out
-			visited = null;
-			ConflictError error;
-			auto conflict_index = validateConfigs(root, error);
-			if (first_error.isNull) first_error = error;
-
-			// print out current iteration state
-			logDebug("Interation (ci=%s) %s", conflict_index, {
-				import std.array : join;
-				auto cs = new string[configs.length];
-				foreach (p, i; package_indices) {
-					if (configs[i].allConfigs.length)
-						cs[i] = p~" "~configs[i].allConfigs[config_indices[i]].to!string~(i >= 0 && i >= conflict_index ? " (C)" : "");
-					else cs[i] = p ~ " [no config]";
+	private void constrainDependencies(TreeNode n, TreeNodes[] dependencies, size_t depidx, ref PackageConfigs[string] packs, ref long loop_counter)
```

Comment:
is the n argument even used in this function?

Reply:
Now it's used in the assertion message at the bottom of the function ;-)